language_model/tensorflow/bert/run_pretraining.py

```diff
@@ -392,24 +396,33 @@ def input_fn(params, input_context = None):
     # For eval, we want no shuffling and parallel reading doesn't matter.
     if is_training:
       d = tf.data.Dataset.from_tensor_slices(tf.constant(input_files))
+      input_length = len(input_files)
       if input_context:
+        input_length = int(
+            math.ceil(input_length / input_context.num_input_pipelines))
         tf.logging.info(
             'Sharding the dataset: input_pipeline_id=%d num_input_pipelines=%d' % (
             input_context.input_pipeline_id, input_context.num_input_pipelines))
         d = d.shard(input_context.num_input_pipelines,
                     input_context.input_pipeline_id)
-      d = d.shuffle(buffer_size=len(input_files))
+      d = d.shuffle(buffer_size=input_length)
 
       # `cycle_length` is the number of parallel files that get read.
-      cycle_length = min(num_cpu_threads, len(input_files))
+      # cycle_length = min(num_cpu_threads, len(input_files))
+      cycle_length = 10
 
       # `sloppy` mode means that the interleaving is not exact. This adds
       # even more randomness to the training pipeline.
-      d = d.apply(
-          tf.data.experimental.parallel_interleave(
-              tf.data.TFRecordDataset,
-              sloppy=is_training,
-              cycle_length=cycle_length))
+      # d = d.apply(
+      #     tf.data.experimental.parallel_interleave(
+      #         tf.data.TFRecordDataset,
+      #         sloppy=is_training,
+      #         cycle_length=cycle_length))
+      d = d.interleave(
+          tf.data.TFRecordDataset,
+          deterministic=False,
+          cycle_length=cycle_length,
+          num_parallel_calls=tf.data.AUTOTUNE)
```

Comment:
@sgpyc What is the purpose of the dataloading changes, and specifically, how does this address the true scale-vs-GA differences that is the core issue of this PR?